// antd@4.24.8/es/pagination/Pagination.js
import _defineProperty from "../../../../../@babel_runtime@7.23.6/node_modules/@babel/runtime/helpers/esm/defineProperty.js";
import _extends from "../../../../../@babel_runtime@7.23.6/node_modules/@babel/runtime/helpers/esm/extends.js";
import DoubleLeftOutlined from "../../../../../@ant-design_icons@4.8.1_react-dom@18.2.0_react@18.2.0/node_modules/@ant-design/icons/es/icons/DoubleLeftOutlined.js";
import DoubleRightOutlined from "../../../../../@ant-design_icons@4.8.1_react-dom@18.2.0_react@18.2.0/node_modules/@ant-design/icons/es/icons/DoubleRightOutlined.js";
import LeftOutlined from "../../../../../@ant-design_icons@4.8.1_react-dom@18.2.0_react@18.2.0/node_modules/@ant-design/icons/es/icons/LeftOutlined.js";
import RightOutlined from "../../../../../@ant-design_icons@4.8.1_react-dom@18.2.0_react@18.2.0/node_modules/@ant-design/icons/es/icons/RightOutlined.js";
import classNames from "../../../../../classnames@2.3.2/node_modules/classnames/index.js";
import Pagination$2 from "../../../../../rc-pagination@3.2.0_react-dom@18.2.0_react@18.2.0/node_modules/rc-pagination/es/Pagination.js";
import enUS from "../../../../../rc-pagination@3.2.0_react-dom@18.2.0_react@18.2.0/node_modules/rc-pagination/es/locale/en_US.js";
import * as React from "react";
import "../../../../../rc-field-form@1.27.4_react-dom@18.2.0_react@18.2.0/node_modules/rc-field-form/es/index.js";
import LocaleReceiver from "../locale-provider/LocaleReceiver.js";
import { ConfigContext } from "../config-provider/context.js";
import useBreakpoint from "../grid/hooks/useBreakpoint.js";
import { MiniSelect, MiddleSelect } from "./Select.js";
var __rest = function(s, e) {
  var t = {};
  for (var p in s)
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
      t[p] = s[p];
  if (s != null && typeof Object.getOwnPropertySymbols === "function")
    for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
      if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
        t[p[i]] = s[p[i]];
    }
  return t;
};
var Pagination = function Pagination2(_a) {
  var customizePrefixCls = _a.prefixCls, customizeSelectPrefixCls = _a.selectPrefixCls, className = _a.className, size = _a.size, customLocale = _a.locale, selectComponentClass = _a.selectComponentClass, responsive = _a.responsive, showSizeChanger = _a.showSizeChanger, restProps = __rest(_a, ["prefixCls", "selectPrefixCls", "className", "size", "locale", "selectComponentClass", "responsive", "showSizeChanger"]);
  var _useBreakpoint = useBreakpoint(responsive), xs = _useBreakpoint.xs;
  var _React$useContext = React.useContext(ConfigContext), getPrefixCls = _React$useContext.getPrefixCls, direction = _React$useContext.direction, _React$useContext$pag = _React$useContext.pagination, pagination = _React$useContext$pag === void 0 ? {} : _React$useContext$pag;
  var prefixCls = getPrefixCls("pagination", customizePrefixCls);
  var mergedShowSizeChanger = showSizeChanger !== null && showSizeChanger !== void 0 ? showSizeChanger : pagination.showSizeChanger;
  var getIconsProps = function getIconsProps2() {
    var ellipsis = /* @__PURE__ */ React.createElement("span", {
      className: "".concat(prefixCls, "-item-ellipsis")
    }, "•••");
    var prevIcon = /* @__PURE__ */ React.createElement("button", {
      className: "".concat(prefixCls, "-item-link"),
      type: "button",
      tabIndex: -1
    }, /* @__PURE__ */ React.createElement(LeftOutlined, null));
    var nextIcon = /* @__PURE__ */ React.createElement("button", {
      className: "".concat(prefixCls, "-item-link"),
      type: "button",
      tabIndex: -1
    }, /* @__PURE__ */ React.createElement(RightOutlined, null));
    var jumpPrevIcon = /* @__PURE__ */ React.createElement("a", {
      className: "".concat(prefixCls, "-item-link")
    }, /* @__PURE__ */ React.createElement("div", {
      className: "".concat(prefixCls, "-item-container")
    }, /* @__PURE__ */ React.createElement(DoubleLeftOutlined, {
      className: "".concat(prefixCls, "-item-link-icon")
    }), ellipsis));
    var jumpNextIcon = /* @__PURE__ */ React.createElement("a", {
      className: "".concat(prefixCls, "-item-link")
    }, /* @__PURE__ */ React.createElement("div", {
      className: "".concat(prefixCls, "-item-container")
    }, /* @__PURE__ */ React.createElement(DoubleRightOutlined, {
      className: "".concat(prefixCls, "-item-link-icon")
    }), ellipsis));
    if (direction === "rtl") {
      var _ref = [nextIcon, prevIcon];
      prevIcon = _ref[0];
      nextIcon = _ref[1];
      var _ref2 = [jumpNextIcon, jumpPrevIcon];
      jumpPrevIcon = _ref2[0];
      jumpNextIcon = _ref2[1];
    }
    return {
      prevIcon,
      nextIcon,
      jumpPrevIcon,
      jumpNextIcon
    };
  };
  return /* @__PURE__ */ React.createElement(LocaleReceiver, {
    componentName: "Pagination",
    defaultLocale: enUS
  }, function(contextLocale) {
    var _classNames;
    var locale = _extends(_extends({}, contextLocale), customLocale);
    var isSmall = size === "small" || !!(xs && !size && responsive);
    var selectPrefixCls = getPrefixCls("select", customizeSelectPrefixCls);
    var extendedClassName = classNames((_classNames = {}, _defineProperty(_classNames, "".concat(prefixCls, "-mini"), isSmall), _defineProperty(_classNames, "".concat(prefixCls, "-rtl"), direction === "rtl"), _classNames), className);
    return /* @__PURE__ */ React.createElement(Pagination$2, _extends({}, getIconsProps(), restProps, {
      prefixCls,
      selectPrefixCls,
      className: extendedClassName,
      selectComponentClass: selectComponentClass || (isSmall ? MiniSelect : MiddleSelect),
      locale,
      showSizeChanger: mergedShowSizeChanger
    }));
  });
};
const Pagination$1 = Pagination;
export {
  Pagination$1 as default
};
