// antd@4.24.8/es/layout/layout.js
import _toConsumableArray from "../../../../../@babel_runtime@7.23.6/node_modules/@babel/runtime/helpers/esm/toConsumableArray.js";
import _defineProperty from "../../../../../@babel_runtime@7.23.6/node_modules/@babel/runtime/helpers/esm/defineProperty.js";
import _slicedToArray from "../../../../../@babel_runtime@7.23.6/node_modules/@babel/runtime/helpers/esm/slicedToArray.js";
import _extends from "../../../../../@babel_runtime@7.23.6/node_modules/@babel/runtime/helpers/esm/extends.js";
import classNames from "../../../../../classnames@2.3.2/node_modules/classnames/index.js";
import * as React from "react";
import "../../../../../rc-field-form@1.27.4_react-dom@18.2.0_react@18.2.0/node_modules/rc-field-form/es/index.js";
import { ConfigContext } from "../config-provider/context.js";
var __rest = function(s, e) {
  var t = {};
  for (var p in s)
    if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
      t[p] = s[p];
  if (s != null && typeof Object.getOwnPropertySymbols === "function")
    for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
      if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
        t[p[i]] = s[p[i]];
    }
  return t;
};
var LayoutContext = /* @__PURE__ */ React.createContext({
  siderHook: {
    addSider: function addSider() {
      return null;
    },
    removeSider: function removeSider() {
      return null;
    }
  }
});
function generator(_ref) {
  var suffixCls = _ref.suffixCls, tagName = _ref.tagName, displayName = _ref.displayName;
  return function(BasicComponent) {
    var Adapter = /* @__PURE__ */ React.forwardRef(function(props, ref) {
      var _React$useContext = React.useContext(ConfigContext), getPrefixCls = _React$useContext.getPrefixCls;
      var customizePrefixCls = props.prefixCls;
      var prefixCls = getPrefixCls(suffixCls, customizePrefixCls);
      return /* @__PURE__ */ React.createElement(BasicComponent, _extends({
        ref,
        prefixCls,
        tagName
      }, props));
    });
    if (process.env.NODE_ENV !== "production") {
      Adapter.displayName = displayName;
    }
    return Adapter;
  };
}
var Basic = /* @__PURE__ */ React.forwardRef(function(props, ref) {
  var prefixCls = props.prefixCls, className = props.className, children = props.children, tagName = props.tagName, others = __rest(props, ["prefixCls", "className", "children", "tagName"]);
  var classString = classNames(prefixCls, className);
  return /* @__PURE__ */ React.createElement(tagName, _extends(_extends({
    className: classString
  }, others), {
    ref
  }), children);
});
var BasicLayout = /* @__PURE__ */ React.forwardRef(function(props, ref) {
  var _classNames;
  var _React$useContext2 = React.useContext(ConfigContext), direction = _React$useContext2.direction;
  var _React$useState = React.useState([]), _React$useState2 = _slicedToArray(_React$useState, 2), siders = _React$useState2[0], setSiders = _React$useState2[1];
  var prefixCls = props.prefixCls, className = props.className, children = props.children, hasSider = props.hasSider, Tag = props.tagName, others = __rest(props, ["prefixCls", "className", "children", "hasSider", "tagName"]);
  var classString = classNames(prefixCls, (_classNames = {}, _defineProperty(_classNames, "".concat(prefixCls, "-has-sider"), typeof hasSider === "boolean" ? hasSider : siders.length > 0), _defineProperty(_classNames, "".concat(prefixCls, "-rtl"), direction === "rtl"), _classNames), className);
  var contextValue = React.useMemo(function() {
    return {
      siderHook: {
        addSider: function addSider2(id) {
          setSiders(function(prev) {
            return [].concat(_toConsumableArray(prev), [id]);
          });
        },
        removeSider: function removeSider2(id) {
          setSiders(function(prev) {
            return prev.filter(function(currentId) {
              return currentId !== id;
            });
          });
        }
      }
    };
  }, []);
  return /* @__PURE__ */ React.createElement(LayoutContext.Provider, {
    value: contextValue
  }, /* @__PURE__ */ React.createElement(Tag, _extends({
    ref,
    className: classString
  }, others), children));
});
generator({
  suffixCls: "layout",
  tagName: "section",
  displayName: "Layout"
})(BasicLayout);
generator({
  suffixCls: "layout-header",
  tagName: "header",
  displayName: "Header"
})(Basic);
generator({
  suffixCls: "layout-footer",
  tagName: "footer",
  displayName: "Footer"
})(Basic);
generator({
  suffixCls: "layout-content",
  tagName: "main",
  displayName: "Content"
})(Basic);
export {
  LayoutContext
};
